Optional features not respected with external registry

I am trying to use a crate that depends on `trust-dns-proto` which has tokio listed as an optional dependency: https://github.com/bluejekyll/trust-dns/blob/0.19.4/crates/proto/Cargo.toml#L77 `
tokio = { version = "0.2.16", optional = true }`

When compiling a simple test project with the Cargo.toml file below, it works as expected as long as I am using the standard Rust crates.io registry.  However, when I try to use an external file system registry created by https://github.com/ChrisGreenaway/cargo-local-registry I receive a compile error. 
Error example:
```
error[E0433]: failed to resolve: use of undeclared type or module `tokio`
  --> /home/mauldinl@gvl.is.l-3com.com/.cargo/registry/src/-7c1c43276983337d/trust-dns-proto-0.19.4/src/lib.rs:25:5
   |
25 | use tokio::runtime::Runtime;
   |     ^^^^^ use of undeclared type or module `tokio`
```

I ran the `cargo tree` command and have attached the two output files.  As you can see, the `tree_output_cratesio.txt` file listed the `tokio` crate as a dependency of `trust-dns-proto` while the `tree_output_external_reg.txt` file does not list the `tokio` crate.  I have tried multiple variations in my `Cargo.toml` file to make sure the `tokio-runtime`  feature is selected and from what I can tell, it is being selected since the line of code that imports the `tokio::runtime::Runtime` line is being executed but for some reason when using an external registry the optional tokio package is not being added in.
I have used the external registry for several months and use other crates that have optional dependencies on other crates like `serde`, etc.. and those work as expected.  I am not sure why I am seeing this behavior. 

### Meta
`rustc --version --verbose`:
rustc 1.44.0 (49cae5576 2020-06-01)
binary: rustc
commit-hash: 49cae55760da0a43428eba73abcb659bb70cf2e4
commit-date: 2020-06-01
host: x86_64-unknown-linux-gnu
release: 1.44.0
LLVM version: 9.0

Cargo.toml example:
```
[package]
name = "tester"
version = "0.1.0"
edition = "2018"

[dependencies]
tokio = {version = "0.2.21", features = ["full"]}
trust-dns-proto = {version = "=0.19.4", default-features = false, features = ["tokio-runtime"]}
```

[tree_output_cratesio.txt](https://github.com/rust-lang/rust/files/4754352/tree_output_cratesio.txt)
[tree_output_external_reg.txt](https://github.com/rust-lang/rust/files/4754353/tree_output_external_reg.txt)

--- a/src/cargo/core/resolver/dep_cache.rs
+++ b/src/cargo/core/resolver/dep_cache.rs
@@ -402,15 +402,13 @@ impl Requirements<'_> {
         // If `package` is indeed an optional dependency then we activate the
         // feature named `package`, but otherwise if `package` is a required
         // dependency then there's no feature associated with it.
-        if let Some(dep) = self
+        if self
             .summary
             .dependencies()
             .iter()
-            .find(|p| p.name_in_toml() == package)
+            .any(|dep| dep.name_in_toml() == package && dep.is_optional())
         {
-            if dep.is_optional() {
-                self.used.insert(package);
-            }
+            self.used.insert(package);
         }
         self.deps
             .entry(package)

--- a/tests/testsuite/features.rs
+++ b/tests/testsuite/features.rs
@@ -1,7 +1,7 @@
 //! Tests for `[features]` table.
 
 use cargo_test_support::paths::CargoPathExt;
-use cargo_test_support::registry::Package;
+use cargo_test_support::registry::{Dependency, Package};
 use cargo_test_support::{basic_manifest, project};
 
 #[cargo_test]
@@ -2112,4 +2112,81 @@ fn slash_optional_enables() {
         .run();
 
     p.cargo("check --features dep/feat").run();
+}
+
+#[cargo_test]
+fn registry_summary_order_doesnt_matter() {
+    // Checks for an issue where the resolver depended on the order of entries
+    // in the registry summary. If there was a non-optional dev-dependency
+    // that appeared before an optional normal dependency, then the resolver
+    // would not activate the optional dependency with a pkg/featname feature
+    // syntax.
+    Package::new("dep", "0.1.0")
+        .feature("feat1", &[])
+        .file(
+            "src/lib.rs",
+            r#"
+                #[cfg(feature="feat1")]
+                pub fn work() {
+                    println!("it works");
+                }
+            "#,
+        )
+        .publish();
+    Package::new("bar", "0.1.0")
+        .feature("bar_feat", &["dep/feat1"])
+        .add_dep(Dependency::new("dep", "0.1.0").dev())
+        .add_dep(Dependency::new("dep", "0.1.0").optional(true))
+        .file(
+            "src/lib.rs",
+            r#"
+                // This will fail to compile without `dep` optional dep activated.
+                extern crate dep;
+
+                pub fn doit() {
+                    dep::work();
+                }
+            "#,
+        )
+        .publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.1.0"
+                edition = "2018"
+
+                [dependencies]
+                bar = { version="0.1", features = ["bar_feat"] }
+            "#,
+        )
+        .file(
+            "src/main.rs",
+            r#"
+                fn main() {
+                    bar::doit();
+                }
+            "#,
+        )
+        .build();
+
+    p.cargo("run")
+        .with_stderr(
+            "\
+[UPDATING] [..]
+[DOWNLOADING] crates ...
+[DOWNLOADED] [..]
+[DOWNLOADED] [..]
+[COMPILING] dep v0.1.0
+[COMPILING] bar v0.1.0
+[COMPILING] foo v0.1.0 [..]
+[FINISHED] [..]
+[RUNNING] `target/debug/foo[EXE]`
+",
+        )
+        .with_stdout("it works")
+        .run();
 }

